### PR TITLE
chore: RFC to update Vector configuration via Helm

### LINKED ISSUE #7709: RFC: Helm chart config rework
While doing work to deprecate the `rawConfig` keys in our charts I realized we have some added complexity and duplication of code for the `managed.toml` configmap.

The purpose of this RFC is to propose a deprecation of existing configuration values and a simplification of our chart templating.

### PR DESCRIPTION
**Pull request by spencergilbert**
Closes #7709

[Rendered](https://github.com/timberio/vector/blob/rfc-update-config-pattern/rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md)

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice work on this. I’m 👍 on the general sentiment of removing the shortcuts and having one place to configure vector.

**Review by jszwedko — COMMENTED** [review]
Nice! This seems more direct and understandable than the special keys.

Regarding compatibility, did you have thoughts? Would we be able to maintain the current behavior at least for a release, but preferably until 1.0 if it isn't a huge burden.

**Comment by spencergilbert** [comment]
> Nice! This seems more direct and understandable than the special keys.
> 
> Regarding compatibility, did you have thoughts? Would we be able to maintain the current behavior at least for a release, but preferably until 1.0 if it isn't a huge burden.

Yes - I think it should actually be super easy. I was toying here was converting everything from the old style into the new [here](https://github.com/timberio/vector/blob/helm-datadog-style/distribution/helm/vector-aggregator/templates/configmap.yaml), but it would be easier to ignore the new style if the old is used and throw a `fail` if both are used.

**Comment by jszwedko** [comment]
> > Nice! This seems more direct and understandable than the special keys.
> > Regarding compatibility, did you have thoughts? Would we be able to maintain the current behavior at least for a release, but preferably until 1.0 if it isn't a huge burden.
> 
> Yes - I think it should actually be super easy. I was toying here was converting everything from the old style into the new [here](https://github.com/timberio/vector/blob/helm-datadog-style/distribution/helm/vector-aggregator/templates/configmap.yaml), but it would be easier to ignore the new style if the old is used and throw a `fail` if both are used.

👍 Do you mind adding those notes about maintaining compatibility to the RFC itself? Otherwise I think this looks good!

**Comment by spencergilbert** [comment]
> 👍 Do you mind adding those notes about maintaining compatibility to the RFC itself? Otherwise I think this looks good!

You mean more fully explained than the next steps section?

**Comment by jszwedko** [comment]
Ah I see it down in "Plan of Attack". I might mention it in the proposal itself since currently it reads like we wouldn't support the old keys due to:

> Ignoring backward compatibility with existing keys, the vector-agent ConfigMap template would be replaced with the following:

**Review by jszwedko — APPROVED** [review]
Nice work. I appreciate you digging into this problem and iterating on the solution to get to this point. I like where you've ended up.

File: rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+today does not mention how to configure any of the default components.
+
+## Rationale
+
+- Provides the easiest mechanism for users to bring their own configuration and disable our provided default config
+- The entire vector configuration will be in a single location for better readability and debuggability
+  - We remove a source of bugs by removing any merging of configuration options (between default provided and user provided)
+- Reduces the templating logic for the entire configuration file to an `if/else` statement and a one-liner to template the provided configuration (if needed)
+- By passing the `customConfig` through a `tpl` function users can more easily generate parts of their configuration from other values (ports, volumes, etc)
+- Matches the [Datadog Agent configuration](https://github.com/DataDog/helm-charts/blob/master/charts/datadog/values.yaml#L1023-L1048)
```
Comment (prognant) [inline rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md:136]: There is also small amount of logic in the config map template https://github.com/DataDog/helm-charts/blob/master/charts/datadog/templates/datadog-yaml-configmap.yaml but overall I think that a good strategy for flexible configuration.

**Review by prognant — APPROVED** [review]
Not much to add, based on the datadog chart usages I've seen, I think this is a convenient pattern, and apart from compatibility issue I don't think there is much to say against doing that.

File: rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md (hunk 1 above)
Comment (spencergilbert) [reply rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md:136]: @prognant were you talking about any specific templating in the configmap? We may include some to ensure ports match configuration and such but I don't think we'll need much.

**Review by StephenWakely — APPROVED** [review]
I think personally I prefer the alternative suggestion here as it just feels a little hacky to rely on uncommenting code to provide defaults. It then means we lose control over updating the defaults for people who only really needed to update a single field should we find more sensible values down the road.

But conversely it means we can update the defaults and worry less about breaking other peoples configsp There seem to be advantages to both ways, so I am not too strongly fixated on either.

Plus this way brings us in line with DataDog so makes it more consistent for folks using the whole ecosystem.

**Pull request merged**